### PR TITLE
125 get undulator gap

### LINKED ISSUE #125: Undulator gap: Get using ophyd
Currently the undulator gap is passed from GDA to Artemis. There is no reason that Artemis could not read the undulator gap itself, this would be good as it will make Artemis more standalone and less dependent on GDA.

## Acceptance Criteria
* There is an Ophyd device for the undulator
* Artemis reads the gap from this device and GDA no longer sends it

### PR DESCRIPTION
**Pull request by MattPrit**
Fixes #125

Undulator device tested using:
```
from bluesky import RunEngine
import bluesky.plan_stubs as bps

undulator = Undulator(name="Undulator", prefix="SR03I-MO-SERVC-01:")

def get_value():
    gap = yield from bps.rd(undulator.gap)
    print(gap)

RE = Runengine({})
RE(get_value())
```
Output was compared with the relevant PV using `caget SR03I-MO-SERVC-01:BLGAPMTR.RBV` and found to be consistent.

Changes needed on the GDA side:
- No longer pass undulator_gap parameter to artemis
- Need to pass insertion_prefix parameter to artemis

### CODE REVIEW

File: src/artemis/fast_grid_scan_plan.py [hunk 1]
```diff
@@ -45,7 +52,7 @@ def run_gridscan(
 
     for id in datacollection_ids:
         run_start(id)
-
+
```
Comment (DominicOram) [inline src/artemis/fast_grid_scan_plan.py:55]: **nit**: Can you remove the superfluous whitespace? I believe vscode should do this for you when you save the file though

**Review by abbiemery — CHANGES_REQUESTED** [review]
Change the name of the test then should be good to go.

File: src/artemis/fast_grid_scan_plan.py [hunk 2]
```diff
@@ -91,7 +98,9 @@ def get_plan(parameters: FullParameters):
     )
     zebra = Zebra(name="zebra", prefix=f"{parameters.beamline}-EA-ZEBRA-01:")
 
-    return run_gridscan(fast_grid_scan, zebra, eiger, parameters)
+    undulator = Undulator(name="undulator", prefix=f"{parameters.beamline}-MO-SERVC-01:")
```
Comment (DominicOram) [inline src/artemis/fast_grid_scan_plan.py:101]: **must**: This isn't quite going to work. The beamline prefix is `BL03I` but the prefix for the undulator is `SR03I`. I think to solve this we need a new parameter called something like `insertion_prefix` that GDA will pass in

File: src/artemis/tests/test_fast_grid_scan_plan.py [hunk 3]
```diff
@@ -23,3 +25,14 @@ def test_given_full_parameters_dict_when_detector_name_used_and_converted_then_d
 def test_when_get_plan_called_then_generator_returned():
     plan = get_plan(FullParameters())
     assert isinstance(plan, types.GeneratorType)
+
+
+def test_parameters_updated_from_epics_devices_correctly():
```
[the remark this answers is not in the document]
Comment (MattPrit) [reply src/artemis/tests/test_fast_grid_scan_plan.py]: Agreed. A more general test can be added later.

File: src/artemis/fast_grid_scan_plan.py (hunk 2 above)
Comment (MattPrit) [reply src/artemis/fast_grid_scan_plan.py]: `beamline` replaced with new parameter `insertion_prefix`

File: src/artemis/fast_grid_scan_plan.py (hunk 1 above)
Comment (MattPrit) [reply src/artemis/fast_grid_scan_plan.py]: yes

**Comment by abbiemery** [comment]
@DominicOram Is it okay in this project to merge PRs that have codecov failing?

**Comment by DominicOram** [comment]
> @DominicOram Is it okay in this project to merge PRs that have codecov failing?

Yes, you improved the code coverage with this PR so it's fine. We have some other tickets to get the general coverage up to the target.

File: src/artemis/fast_grid_scan_plan.py [hunk 4]
```diff
@@ -91,7 +102,11 @@ def get_plan(parameters: FullParameters):
     )
     zebra = Zebra(name="zebra", prefix=f"{parameters.beamline}-EA-ZEBRA-01:")
 
-    return run_gridscan(fast_grid_scan, zebra, eiger, parameters)
+    undulator = Undulator(
+        name="undulator", prefix=f"{parameters.beamline}-MO-SERVC-01:"
```
Comment (DominicOram) [inline src/artemis/fast_grid_scan_plan.py:106]: **must**: This is still using the beamline prefix rather than the insertion prefix.

**Review by DominicOram — CHANGES_REQUESTED** [review]
Thanks for adding the new prefix but we still need to use it for the Undulator...

File: src/artemis/fast_grid_scan_plan.py (hunk 4 above)
Comment (MattPrit) [reply src/artemis/fast_grid_scan_plan.py]: Whoops. Fixed.

**Review by DominicOram — APPROVED** [review]
Looks good to me, thanks! Feel free to merge, I think you've fixed @abbiemery's comment.

**Pull request merged**